doc/decisions/reference_counting.md

```diff
@@ -0,0 +1,39 @@
+# Reference Counting
+
+## Problem
+
+- locking is not reset when ref counting again gets 0 (adding to keyset and
+  pop again) #2202
+- C++ API for KeySet and Key has unexpected differences: also use ref counting
+  for KeySets (also suggested in #1332)
+
+## Constraints
+
+## Assumptions
+
+## Considered Alternatives
+
+- make ref counting thread safe (probably useful for JNI)
+- start with 1 for reference counting and let keyDecRef do keyDel
+
+## Decision
+
+- add second reference counter to Key (size 1 byte)
```

Comment:
```suggestion
- add second reference counter to Key
```

I think the size is still TBD. Also it shouldn't really be something that is fixed in the decision.